gst/nnstreamer/tensor_query/nnstreamer_edge.h

```diff
@@ -0,0 +1,166 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * Copyright (C) 2022 Gichan Jang <gichan2.jang@samsung.com>
+ *
+ * @file   nnstreamer_edge.h
+ * @date   25 Mar 2022
+ * @brief  Common library to support communication among devices.
+ * @see    https://github.com/nnstreamer/nnstreamer
+ * @author Gichan Jang <gichan2.jang@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ * @note This file will be moved to nnstreamer-edge repo. (https://github.com/nnstreamer/nnstreamer-edge)
+ * Need update license when migrating this into edge repo.
+ */
+
+#ifndef __NNSTREAMER_EDGE_H__
+#define __NNSTREAMER_EDGE_H__
+
+#include <stddef.h>
+#include <stdbool.h>
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+/**
+ * @brief The maximum number of data instances that nnstreamer-edge data may have.
+ */
+#define NNS_EDGE_DATA_LIMIT (16)
+
+typedef void *nns_edge_h;
+typedef void *nns_edge_event_h;
+typedef void *nns_edge_data_h;
+
+/**
+ * @brief Enumeration for the error codes of nnstreamer-edge.
+ * @todo define error code later (linux standard error code)
+ */
+typedef enum
+{
+  NNS_EDGE_ERROR_NONE = 0,
+  NNS_EDGE_ERROR_UNKNOWN = -1,
+  NNS_EDGE_ERROR_INVALID_PARAMETER = -2,
+  NNS_EDGE_ERROR_OUT_OF_MEMORY = -3,
+  NNS_EDGE_ERROR_CONNECTION_FAILURE = -4,
+  NNS_EDGE_ERROR_IO = -5,
+} nns_edge_error_e;
+
+typedef enum
+{
+  NNS_EDGE_EVENT_TEMP = 0,
+
+  NNS_EDGE_EVENT_MAX
+} nns_edge_event_e;
+
+typedef enum
+{
+  NNS_EDGE_PROTOCOL_TCP = 0,
+  NNS_EDGE_PROTOCOL_MQTT,
+  NNS_EDGE_PROTOCOL_AITT,
+  NNS_EDGE_PROTOCOL_AITT_TCP,
+
+  NNS_EDGE_PROTOCOL_MAX
+} nns_edge_protocol_e;
+
+typedef enum
+{
+  NNS_EDGE_DATA_TYPE_INFO = 0,
+  NNS_EDGE_DATA_TYPE_RAW_DATA,
+
+  NNS_EDGE_DATA_TYPE_MAX
+} nns_edge_data_type_e;
+
+/**
+ * @brief Callback for the nnstreamer edge event.
+ */
+typedef int (*nns_edge_event_cb) (nns_edge_event_h event_h, void *user_data);
+
+/**
+ * @brief Callback called when nnstreamer-edge data is released.
+ */
+typedef void (*nns_edge_data_destroy_cb) (void *data);
+
+/**
+ * @brief Get registered handle. If not registered, create new handle and register it.
+ */
+int
+nns_edge_get_handle (const char *id, const char *topic, nns_edge_h *edge_h);
```

Comment:
The name of Edge APIs is intuitive so easy to make a code. However, there is no example to refer to.
What about adding some pseudo-code for API usage? In my opinion, simple code without error checking is good enough as below.

```cpp
nns_edge_h handle;
nns_edge_data_h data_h;

/* create the edge handle and connect the MQTT broker */
nns_edge_get_handle (USER_ID, "object_detect/mobile_v3", &handle);
nns_edge_connect (handle, NNS_EDGE_PROTOCOL_MQTT, BROKER_IP, BROKER_PORT);

/* create the edge data to publish */
nns_edge_data_create (NNS_EDGE_DATA_TYPE_RAW_DATA, &data_h);
nns_edge_data_add (data_h, buffer, buffer_size, NULL);

/* publish single MQTT message */
nns_edge_publish (handle, data_h);

/* finalize handles */
nns_edge_data_destroy (data_h);
nns_edge_disconnect (handle);
nns_edge_release_handle (handle);
```

Reply:
marked todo (add sample code)

Reply:
Verified!